### ISSUE TITLE
is very hard to factor products of two large prime numbers

### ISSUE THREAD

**Issue by radumarias** [radumarias/Pavel_Ruban_interview#2]
https://github.com/radumarias/Pavel_Ruban_interview/blob/c0601a86f94e796d48964abddd9997e6be800322/problem_5_bonus/src/main/java/com/test/problem5/service/impl/CryptoPangramService.java#L64

in solution you factorize which means you didn't understood the main idea of this problem, which also is the basis of Public-key cryptography (all what you are using with SSL which is at the core of any HTTPS site you are visiting, almost all ideally :) )  which is stated in problem specs also, that “is very hard to factor products of two large prime numbers” :)
> we learned that it is very hard to factor products of two large prime numbers, so we devised an encryption scheme based on that fact.
  The solution is very inefficient as it has a big run complexity order (Big O notation) with many nested for statements and seems not even correct as for case 2 it doesn’t work and most likely for more complex cases.

**Comment by PavelRuban** [comment]
but what is the problem with even value 2102 in case 2? I don't understand why it's there, as soon as only odd primes used to product cipher. how can it be?

**Comment by PavelRuban** [comment]
also there is one more even value 3722 in the cipher case 2

**Comment by radumarias** [comment]
> but what is the problem with even value 2102 in case 2? I don't understand why it's there, as soon as only odd primes used to product cipher. how can it be?
> btw about P5. test case 2 contains value 2102, which is even. it contradicts with the problem description, as soon as only odd primes should be used. so it looks like typo imho.
- no, because in specs they say "we chose to use the first 26 odd prime numbers", but you get the ciphertext as input, which for sure are not the primes :)
> we chose to use the first 26 odd prime numbers, because we worry that it is too easy to factor even numbers

**Comment by radumarias** [comment]
if you would get primes as input then encryption is useless :)

**Comment by PavelRuban** [comment]
I said that odd primes can't give us even cipher values

**Comment by PavelRuban** [comment]
I didn't said we got primes as input

**Comment by PavelRuban** [comment]
![2020-09-02 21_31_36-Cryptopangrams pdf - Foxit Reader](https://user-images.githubusercontent.com/2482396/92022497-03885400-ed64-11ea-8f85-001ff2a2c244.png)
what is that?

anyway, I'm agree solution is inefficient, I know it.

**Comment by radumarias** [comment]
how they modeled it can, because there is a catch there on how they compose the ciphertext and that's they key on how you can solve it efficiently without factorize

**Comment by radumarias** [comment]
don't get me wrong, it's ok how you tried, just that is by far the solution for this problem. the whole point was not to factorize as the specs asks also :)

**Comment by PavelRuban** [comment]
no, I think my try is not ok. I have to find greatest common divisor instead of factorizing. euclid's algorithm. my bad :(

**Comment by PavelRuban** [comment]
well I can implement it during nearest hour (max)
sorry for wasting your time -_-

**Comment by radumarias** [comment]
> no, I think my try is not ok. I have to find greatest common divisor instead of factorizing. euclid's algorithm. my bad :(

that's the catch for the solution :)

**Comment by PavelRuban** [comment]
yeah, I didn't spend enough time and energy for this problem and forgot basic math skills.. :) and I thank you a lot for make me remember part of these skills. unfortunately I hadn't faced math-related problems often enough to keep my mind in appropriate condition )

**Comment by radumarias** [comment]
Relax, this is a hard problem and yes unfortunately we endup needing to work in all kind of features and bugs and we don't have time to implement nice stuff so much. Believe me I miss also a lot of challenges with beautiful algorithms and times like tests and other public contests are great for this :)
